warning: _FORTIFY_SOURCE requires compiling with optimization (-O)

Debug build on Fedora 21:

	$ g++ --version
	g++ (GCC) 4.9.2 20150212 (Red Hat 4.9.2-6)
	Copyright (C) 2014 Free Software Foundation, Inc.
	This is free software; see the source for copying conditions.  There is NO
	warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

	$ ./configure --enable-debug
	[...]
	checking whether compiler predefines _FORTIFY_SOURCE... no
	[...]

	$ make V=1
	[...]
	g++ -I./../include -I../include    -std=c++11 -D_FORTIFY_SOURCE=2 -ggdb3 -Wall -W -Wno-unused-parameter -Woverloaded-virtual -Wshadow -pthread -fvisibility=hidden -fPIC -include znc/zncconfig.h -c -o autocycle.o autocycle.cpp  -MD -MF .depend/autocycle.o.dep
	In file included from /usr/include/c++/4.9.2/x86_64-redhat-linux/bits/os_defines.h:39:0,
	                 from /usr/include/c++/4.9.2/x86_64-redhat-linux/bits/c++config.h:2109,
	                 from /usr/include/c++/4.9.2/bits/stl_algobase.h:59,
	                 from /usr/include/c++/4.9.2/bits/stl_tree.h:61,
	                 from /usr/include/c++/4.9.2/map:60,
	                 from ./../include/znc/ZNCString.h:21,
	                 from ./../include/znc/Nick.h:21,
	                 from ./../include/znc/Chan.h:21,
	                 from autocycle.cpp:17:
	/usr/include/features.h:328:4: warning: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Wcpp]
	 #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
	    ^
	[...]


<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8884444-warning-_fortify_source-requires-compiling-with-optimization-o?utm_campaign=plugin&utm_content=tracker%2F1759&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1759&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/configure.ac
+++ b/configure.ac
@@ -228,24 +228,6 @@ case "${host_os}" in
 	;;
 esac
 
-# In old times needed to define _FORTIFY_SOURCE to 2 ourself.
-# Then GCC started to define it itself to 2. It was ok.
-# But then GCC 4.7 started to define it to 0 or 2 depending on optimization level, and it started to conflict with our define.
-AC_MSG_CHECKING([whether compiler predefines _FORTIFY_SOURCE])
-AC_COMPILE_IFELSE([
-	AC_LANG_PROGRAM([[
-	]], [[
-		#ifndef _FORTIFY_SOURCE
-		#error "Just checking, nothing fatal here"
-		#endif
-	]])
-], [
-	AC_MSG_RESULT([yes])
-], [
-	AC_MSG_RESULT([no])
-	appendCXX "-D_FORTIFY_SOURCE=2"
-])
-
 if test "$DEBUG" != "no"; then
 	appendCXX -ggdb3
 	AC_DEFINE([_DEBUG], [1], [Define for debugging])
@@ -258,6 +240,24 @@ if test "$DEBUG" != "no"; then
 else
 	if test "x$OPTIMIZE" = "xyes"; then
 		appendCXX -O2
+
+		# In old times needed to define _FORTIFY_SOURCE to 2 ourself.
+		# Then GCC started to define it itself to 2. It was ok.
+		# But then GCC 4.7 started to define it to 0 or 2 depending on optimization level, and it started to conflict with our define.
+		AC_MSG_CHECKING([whether compiler predefines _FORTIFY_SOURCE])
+		AC_COMPILE_IFELSE([
+			AC_LANG_PROGRAM([[
+			]], [[
+				#ifndef _FORTIFY_SOURCE
+				#error "Just checking, nothing fatal here"
+				#endif
+			]])
+		], [
+			AC_MSG_RESULT([yes])
+		], [
+			AC_MSG_RESULT([no])
+			appendCXX "-D_FORTIFY_SOURCE=2"
+		])
 	fi
 fi
 

